src/rust/kafka/src/lib.rs

```diff
@@ -150,6 +157,71 @@ impl<T: SerDe> Producer<T> {
     }
 }
 
+#[derive(Clone)]
+pub struct BytesProducer {
+    producer: FutureProducer,
+    topic: String,
+}
+
+impl BytesProducer {
+    pub fn new(config: ProducerConfig) -> Result<Self, ConfigurationError> {
+        Ok(Self {
+            producer: producer(
+                config.bootstrap_servers,
+                config.sasl_username,
+                config.sasl_password,
+            )?,
+            topic: config.topic,
+        })
+    }
+
+    #[tracing::instrument(err, skip(self))]
+    pub async fn send(&self, msg: Bytes) -> Result<(), ProducerError> {
+        let record: FutureRecord<[u8], [u8]> = FutureRecord::to(&self.topic).payload(&msg);
+
+        self.producer
+            .send(record, Timeout::Never)
+            .map(|res| -> Result<(), ProducerError> {
+                res.map_err(|(e, _)| -> ProducerError { e.into() })
+                    .map(|(partition, offset)| {
+                        tracing::trace!(
+                            message = "wrote message",
+                            partition = partition,
+                            offset = offset,
+                        );
+                    })
+            })
+            .await
+    }
+}
+
+#[derive(Clone)]
+pub struct RetryProducer<T>
+where
+    T: SerDe,
+{
+    producer: Producer<Envelope<T>>,
+}
+
+impl<T: SerDe> RetryProducer<T> {
+    pub fn new(config: RetryProducerConfig) -> Result<Self, ConfigurationError> {
+        Ok(Self {
+            producer: Producer::new(ProducerConfig {
+                bootstrap_servers: config.bootstrap_servers,
+                sasl_username: config.sasl_username,
+                sasl_password: config.sasl_password,
+                topic: config.topic,
+            })?,
+        })
+    }
+
+    #[tracing::instrument(err, skip(self))]
+    pub async fn send(&self, mut msg: Envelope<T>) -> Result<(), ProducerError> {
+        msg.metadata.retry_count += 1;
+        self.producer.send(msg.clone()).await
```

Comment:
The clone here seems unlikely to be necessary

Reply:
fixed